### PR TITLE
gh-126180: Remove getopt and optparse deprecation notices

### LINKED ISSUE #126180: Undeprecate getopt and optparse
See discussion at https://discuss.python.org/t/getopt-and-optparse-vs-argparse/69618. Seems that all support this.

`argparse` has many issues, some are fundamental and cannot be easily worked around. Its implements behavior different from the behavior of other Unix/Linux programs in subtle ways. It is still not mature to be the preferable Python module for CLI.

`getopt` is a de-facto cross-language standard. Many programmers that came from other programming languges are already familiar with it. It is very simple, so even if there are other CLI modules, it should be left in the stdlib.

`optparse` provides an object-oriented interface. Even if it has less features than `argparse`, it is more stable and has less bugs. It should be the preferable Python module for CLI.

### Linked PRs
* gh-126186
* gh-126227
* gh-128191

### PR DESCRIPTION
**Pull request by ncoghlan**
getopt and optparse are in no way at risk of being removed,
as they have genuinely valid use cases. Those use cases
(mimicing the C getopt API, implementing third party
command line argument processing libraries) are just
far less common than the ones that argparse addresses.

Closes #126180.

* Issue: gh-126180

----
📚 Documentation preview 📚: https://cpython-previews--126227.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/optparse.rst
Comment (ncoghlan) [inline Doc/library/optparse.rst:130]: Replacing `old` with `minimalist` is the only actual wording change here (the rest is just reflowing text)

File: Doc/howto/argparse.rst
Comment (serhiy-storchaka) [inline Doc/howto/argparse.rst]: This may create an impression that `optparse` is lower level than `getopt`.

File: Doc/library/argparse.rst
Comment (serhiy-storchaka) [inline Doc/library/argparse.rst]: Not only this. There are other peculiarities that make implementation of compatible interface with `argparse` impossible (unlike to `getopt` and `optparse`).

File: Doc/library/argparse.rst
Comment (serhiy-storchaka) [inline Doc/library/argparse.rst]: I would not recommend it for now. You can use it on your own risk. `optparse` may need more work, but it works as expected. I would recommend it for beginners, which cannot distinguish their own errors from peculiarities of the library.

File: Doc/library/cmdlinelibs.rst
Comment (serhiy-storchaka) [inline Doc/library/cmdlinelibs.rst:16]: I would place `optparse` above `argparse`.

File: Doc/library/getopt.rst
Comment (serhiy-storchaka) [inline Doc/library/getopt.rst:10]: I would convert the note in normal paragraph. Also, it should refer to `optparse` as an object-oriented and extensible alternative, not only to `argparse` (which is not an equivalent).

File: Doc/library/getopt.rst
Comment (serhiy-storchaka) [inline Doc/library/getopt.rst]: Even in this simple example this is not an exact equivalent. For example, `-o -v` and `-o --` do not work, `-o=foo` works differently, you cannot interleave options and positional arguments. I said this is "a roughly equivalent".

File: Doc/library/getopt.rst
Comment (serhiy-storchaka) [inline Doc/library/getopt.rst]: I added this to show the difference in the API between three modules
```py
       parser.add_argument('rest', nargs='*')
       args = parser.parse_args()
       process(args.rest, output=args.output, verbose=args.verbose)
```

File: Doc/library/getopt.rst
Comment (serhiy-storchaka) [inline Doc/library/getopt.rst]: The difference is not only in this.

**Review by serhiy-storchaka — COMMENTED** [review]
You are focused on one particular peculiarity in `argparse`, but there are many others. The difference is fundamental in the way how optional and positional arguments are parsed.

File: Doc/library/argparse.rst
Comment (ncoghlan) [reply Doc/library/argparse.rst]: OK, I had gotten the impression from the Discourse thread that this was the primary quirk, but I can reword these caveats to be less specific to that one issue.

File: Doc/howto/argparse.rst
Comment (ncoghlan) [reply Doc/howto/argparse.rst]: I'll reorder it to make it clear the "lower level" comparison is relative to `argparse`

File: Doc/library/argparse.rst
Comment (ncoghlan) [reply Doc/library/argparse.rst]: For this PR, I'd like to stick with the status quo as far as recommendations go (that is, only taking the step back from "`argparse` is the only non-deprecated argument processing option in the standard library").

Taking the extra step to saying "`argparse` is *not* recommended over `optparse` anymore, even for end user applications" would then be a separate follow up question.

File: Doc/library/getopt.rst
Comment (ncoghlan) [reply Doc/library/getopt.rst:10]: I ended up moving this page back to the "Superseded" section of the docs, as it really is a niche use case when anyone should be reaching for this. The use of the note syntax here reflects that "You almost certainly don't want this module" status.

I agree about referring readers to optparse in the note, though.

File: Doc/library/getopt.rst
Comment (ncoghlan) [reply Doc/library/getopt.rst]: I think if we elaborate on the discrepancies here, it may help lay the foundation for weakening the current recommendation to use argparse in a future PR. I'll incorporate more of your notes here tomorrow.

**Comment by ncoghlan** [comment]
@serhiy-storchaka I think several of your suggestions are good improvements. I should have time to incorporate them tomorrow (Nov 1 AEST)

File: Doc/library/optparse.rst
Comment (LilyFirefly) [inline Doc/library/optparse.rst]: It might make sense to duplicate these `seealso` notes in `Doc/library/argparse.rst` and `Doc/howto/argparse.rst`.

**Comment by ncoghlan** [comment]
@serhiy-storchaka I won't get back to this today, and this weekend isn't looking good either. Early next week will probably be the next update.

File: Doc/library/optparse.rst
Comment (ncoghlan) [reply Doc/library/optparse.rst]: I'm wary of the potential for confusion for folks that are running through "My first command line application" style tutorials, so I'd like to avoid having too much "Should you be using this?" text at the start of the `argparse` docs. 

Having an off-ramp from `argparse` to `optparse` and then from there to PyPI alternatives seems like it could be a reasonable middle ground.

**Comment by ncoghlan** [comment]
@serhiy-storchaka I ended up procrastinating on my assignment by working on this (amongst other things), so I've pushed some updates for you to take a look at when you're back at work.

File: Doc/library/getopt.rst
Comment (ncoghlan) [inline Doc/library/getopt.rst]: @serhiy-storchaka ^^^ is the main reason I don't think we should walk back the default argparse recommendation. Many (probably even most) Python devs would consider argparse to be in the right for these 3 examples, and getopt/optparse the ones doing something dubious (even if what they're doing is the common Unix convention).

The discrepancies are valid reasons for reverting the soft deprecation (optparse can genuinely be used to define command line behaviours that argparse doesn't support), but it will take less arguable examples than these ones to make the case that optparse should be preferred over argparse in general.

File: Doc/library/getopt.rst
Comment (serhiy-storchaka) [inline Doc/library/getopt.rst]: Wrong indentation. Also, add `.. testcode::`.

File: Doc/library/getopt.rst
Comment (serhiy-storchaka) [reply Doc/library/getopt.rst]: It would be better to write about these peculiarities in the `argparse` module documentation. The `getopt` module documentation is not the place where users will search information about the `argparse` module.

**Review by serhiy-storchaka — COMMENTED** [review]
I do not agree with some of these changes, but overall this is a step in the right direction. Let's not drag it in bikeshedding.

File: Doc/howto/argparse-optparse.rst
Comment (mwichmann) [inline Doc/howto/argparse-optparse.rst]: Let me try squinting at this from a slightly different angle, feel free to ignore...

:mod:`argparse` was originally created to address some needs that the design of :mod:`optparse` made difficult to implement. While initially the majority of the :mod:`optparse` interface was imported as a transitional aid, maintaining backwards compatibility was never a goal.

File: Doc/howto/argparse.rst
Comment (mwichmann) [inline Doc/howto/argparse.rst]: Python `getopt` provides behavior that combines the classic `getopt` and the GNU `getopt_long`, so the reference (I realize it was there in the original) is incomplete. And to be a little nitpicky, the ISO C standard says nothing about `getopt` (POSIX does, but only the single-char `getopt`) so From C could be mildly confusing. Maybe something like: _(which specifically serves as an equivalent to the getopt family of functions familiar to C language programmers)_?

File: Doc/library/argparse.rst
Comment (gpshead) [inline Doc/library/argparse.rst]: I'd editorialize a little less and elide "stricter" & "common Unix and Linux" and simplify this to "adherence to command line interface concentions related to the handling of values that start with -"

File: Doc/library/argparse.rst
Comment (gpshead) [reply Doc/library/argparse.rst]: Indeed. Keep the argparse recommendation. We're not going to dis-recommend the most widely used Python argument parsing library in the world (that causes churn and angst from users). Nor are we going to be adding yet another one to the stdlib in the future. argparse supports most users quite well, "warts" and all, with plenty of popular third party options for yet other behaviors beyond our stdlib options.

File: Doc/library/cmdlinelibs.rst
Comment (gpshead) [reply Doc/library/cmdlinelibs.rst:16]: If anything, this list sure be in purely alphabetical order.  passive aggressive opinions via list ordering aren't helpful. 😄

File: Doc/library/getopt.rst
Comment (gpshead) [reply Doc/library/getopt.rst]: agreed. I like that we're giving a list of differences in behavior, but I think it makes more sense to put this in a common section :ref:'d from all three of our command line parsing module docs. I don't expect anyone reading argparse (most likely) or optparse (likely) docs to find this in the getopt docs (presumed least likely to be read).

File: Doc/library/getopt.rst
Comment (merwok) [inline Doc/library/getopt.rst]: IMO the salient feature of optparse over getopt is that it’s more convenient and automated! «more OO» is not a selling point

**Comment by ncoghlan** [comment]
Status update: I don't expect to make any progress this week, but I'll be down at the PyCon AU sprints next week, and this is high on my list of tasks to tackle then.

Some high level thoughts:

I don't want to add a completely new page for the trade-off comparison, but I agree hiding it away at the end of the getopt docs isn't sufficient. The approach I'm planning to try next is to:

* add a dedicated section for it at the start of the *optparse* docs. "Should I use `optparse` over `argparse`?" is a question that genuinely needs answering, since there are valid reasons to do it (hence reversing the deprecation), but I want to keep it out of the general reading path for argparse-based Python CLI app development tutorials
* Reference that new section from the start of the argparse docs, as well as from the getopt docs (in the latter case, primarily when explaining why the argparse version is only approximately equivalent to the getopt version, but also earlier when suggesting that readers consider using optparse or argparse instead)
* do a full "tone" review on the optparse docs, since we never seriously revised them after argparse was added, so there are a few areas where they're still assuming that getopt is the only relevant point of comparison, which hasn't been true for a long time

I'll also aim to tidy up the getopt C references as @mwichmann noted.

**Comment by serhiy-storchaka** [comment]
Can we first make simpler changes, undeprecate getopt and optparse, and continue tinkering the documentation in the following PR?

**Comment by ncoghlan** [comment]
@serhiy-storchaka An intermediate step would definitely be helpful, as I didn't get to this at the PyCon AU sprints, and this weekend isn't looking promising either. It will probably be Dec 7th at the earliest now before I find another decent chunk of time to devote to it.

The scope I would suggest for an interim update would be:

* soften deprecation notices in the `optparse` and `getopt` docs back to something more like their old wording
* remove the deprecation announcements from the 3.13 What's New doc
* fix the misleading example at the end of the `getopt` docs

The `argparse` docs and moving the `optparse` link back out of the "superseded" docs section is best left until the broader follow-up patch, as I think it's once we're touching those that we need to be careful with how any notices of potential problems may be perceived by beginners.

File: Doc/library/getopt.rst
Comment (ncoghlan) [reply Doc/library/getopt.rst]: I've replaced "object-oriented" with "declarative" in the places where I had used it.

File: Doc/library/argparse.rst
Comment (ncoghlan) [reply Doc/library/argparse.rst]: Reworded it to be more explicitly descriptive: "... or accepting option parameter values that start
   with ``-`` even when they correspond to another defined option."

File: Doc/howto/argparse.rst
Comment (ncoghlan) [reply Doc/howto/argparse.rst]: Nice phrasing! I've used this in a couple of places (but with "available" instead of "familiar", since not all C programmers will be familiar with them, but they will almost always have them available)

File: Doc/howto/argparse-optparse.rst
Comment (ncoghlan) [reply Doc/howto/argparse-optparse.rst]: Good catch. I didn't use that exact wording, but I did rewrite this paragraph to refer to the fact that the two modules solve different problems, so their APIs are inevitably different.

File: Doc/library/getopt.rst
Comment (ncoghlan) [reply Doc/library/getopt.rst]: Done (in next push)

File: Doc/library/getopt.rst
Comment (ncoghlan) [reply Doc/library/getopt.rst]: I've moved this to a new "Choosing an argument parsing library" section at the start of the `optparse` docs. `getopt` still has the example code, but refers to that new section to explain the differences.

Similar, I made the note at the start of the `argparse` docs shorter, and added a link to this new section.

File: Doc/library/cmdlinelibs.rst
Comment (ncoghlan) [reply Doc/library/cmdlinelibs.rst:16]: Technically, it's alphabetical within topic areas (argument parsing, secure user input, file input, text UIs), but the categorisation and ordering of the topic areas is pretty arbitrary (in truth, it's just the previous order from the `allos.rst` page, with `optparse` restored to its position after `argparse`, and `fileinput` added before `curses`.

File: Doc/library/getopt.rst
Comment (merwok) [inline Doc/library/getopt.rst]:
```suggestion
   if __name__ == '__main__':
       parser = optparse.OptionParser()
       parser.add_option('-o', '--output')
       parser.add_option('-v', dest='verbose', action='store_true')
       opts, args = parser.parse_args()
       process(args, output=opts.output, verbose=opts.verbose)
```

File: Doc/library/getopt.rst
Comment (merwok) [inline Doc/library/getopt.rst]: This block also uses 6-space indentation

**Comment by ncoghlan** [comment]
I forgot I had marked this as draft when it was clear it was going to take me a while to get back to it. Belatedly switched it back to ready for review after the last round of updates (this is the full version of the patch - once I made time for it, I made enough time for that rather than pursuing the interim change discussed above).

**Comment by ncoghlan** [comment]
The DO-NOT-MERGE label was a mutual one with #126186 until we decided which of the PRs to use as the basis for docs updates. With #126186 closed, that means this one can be handled via the regular PR review process.

File: Doc/library/argparse.rst
Comment (ncoghlan) [reply Doc/library/argparse.rst]: Current text covers more discrepancies (and makes it clear it still isn't an exhaustive list)

File: Doc/library/getopt.rst
Comment (ncoghlan) [reply Doc/library/getopt.rst]: This topic now has a dedicated section at the start of the optparse docs (referenced from the argparse and getopt docs).

File: Doc/library/getopt.rst
Comment (ncoghlan) [reply Doc/library/getopt.rst]: This is now covered at the start of the optparse docs.

File: Doc/library/getopt.rst
Comment (ncoghlan) [reply Doc/library/getopt.rst]: Now covered at the start of the optparse docs.

File: Doc/library/optparse.rst
Comment (ncoghlan) [reply Doc/library/optparse.rst]: I'm happy with the way the docs look with the argparse offramp to PyPI running through the new argument parsing library selection notes at the start of the optparse docs, so sticking with that approach (at least for now - future PRs are always free to proposed adjusting things further).

**Review by gpshead — APPROVED** [review]

**Comment by ncoghlan** [comment]
@serhiy-storchaka Let me know if you want to take another look before I merge this, otherwise I'll merge it early next week (a Christmas present for happy optparse and getopt users!).

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Comment by gpshead** [comment]
From my perspective this is ready to go in. I didn't immediately click merge as I wanted to let others speak up if they believe otherwise. Edited quoting of myself from Discord:

"""
It's not like there is any big official "plan" (re: `getopt`, `optparse`, vs `argparse`) - the point of this PR _to me_ is to remove the loaded term "deprecated" on the old ones because we're never going to _remove_ them. We just want to guide users to `argparse` by default and point at the others as also available for specific needs.  Between `sys.argv`, these three modules, and everything on PyPI I'm sure we have 14 ways to do command line parsing that are all wrong from someones perspective anyways.
I wouldn't _encourage_ outsiders to contribute new features to `getopt` or `optparse` _myself_. But if a core dev is willing to review them and guarantee backwards compatibility with tests, go for it.
""" - me

Feel free to merge when you're ready.

**Comment by ncoghlan** [comment]
Done! Thanks for the reviews, folks.

**Pull request merged**
